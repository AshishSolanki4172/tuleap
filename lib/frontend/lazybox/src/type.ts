/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { HTMLTemplateResult, TemplateResult } from "lit/html.js";
import type { GroupCollection, LazyboxItem } from "./items/GroupCollection";
import type { HTMLTemplateStringProcessor } from "./index";

export interface Lazybox {
    setDropdownContent: (groups: GroupCollection) => void;
    resetSelection: () => void;
    setSelection: (selection: ReadonlyArray<LazyboxItem>) => void;
    destroy: () => void;
}

export type LazyboxSearchFieldCallback = (query: string) => void;

export type LazyboxSelectionCallback = (selected_value: unknown | null) => void;

export type LazyboxTemplatingCallback = (
    html: typeof HTMLTemplateStringProcessor,
    item: LazyboxItem
) => HTMLTemplateResult;

export type LazyboxNewItemCallback = () => void;

export type LazyBoxWithNewItemButton = {
    readonly new_item_button_label: string;
    readonly new_item_callback: LazyboxNewItemCallback;
};

type LazyBoxWithoutNewItemButton = {
    readonly new_item_callback?: undefined;
};

export type LazyboxOptions = (LazyBoxWithoutNewItemButton | LazyBoxWithNewItemButton) & {
    readonly placeholder: string;
    readonly search_input_placeholder: string;
    readonly is_multiple: boolean;
    readonly templating_callback: LazyboxTemplatingCallback;
    readonly selection_callback: LazyboxSelectionCallback;
    readonly search_field_callback: LazyboxSearchFieldCallback;
};

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: unknown;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
}

export interface LazyboxComponent {
    wrapper_element: HTMLElement;
    lazybox_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LazyboxSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface LazyboxSelectionStateMultiple {
    selected_items: Map<string, RenderedItem>;
    selected_values_elements: Map<string, Element>;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}

export interface ManageSelection {
    processSelection(item: Element): void;
    hasSelection(): boolean;
    updateSelectionAfterDropdownContentChange(): void;
    clearSelection(): void;
    setSelection(selection: ReadonlyArray<RenderedItem>): void;
}
