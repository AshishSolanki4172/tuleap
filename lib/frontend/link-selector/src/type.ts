/**
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import type { TemplateResult } from "lit/html.js";
import type { LinkSelectorSearchFieldCallback } from "./events/SearchFieldEventCallbackHandler";

export interface LinkSelector {
    destroy: () => void;
}

export interface LinkSelectorOptions {
    readonly placeholder?: string;
    search_field_callback: LinkSelectorSearchFieldCallback;
}

export type RenderedItemMap = Map<string, RenderedItem>;

export interface RenderedItem {
    id: string;
    template: TemplateResult;
    value: string;
    is_disabled: boolean;
    is_selected: boolean;
    group_id: string;
    element: Element;
    target_option: HTMLOptionElement;
}

export interface LinkSelectorComponent {
    wrapper_element: HTMLElement;
    link_selector_element: Element;
    dropdown_element: HTMLElement;
    selection_element: HTMLElement;
    placeholder_element: Element;
    dropdown_list_element: HTMLElement;
    search_field_element: HTMLInputElement;
}

export interface LinkSelectorSelectionStateSingle {
    selected_item: RenderedItem;
    selected_value_element: DocumentFragment;
}

export interface ScrollCoordinates {
    x_position: number;
    y_position: number;
}
