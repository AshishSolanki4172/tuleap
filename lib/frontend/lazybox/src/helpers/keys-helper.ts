/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

export function isEscapeKey(event: KeyboardEvent): boolean {
    return event.key === "Escape" || event.key === "Esc";
}

export function isEnterKey(event: KeyboardEvent): boolean {
    return event.key === "Enter";
}

export function isArrowDown(event: KeyboardEvent): boolean {
    return event.key === "ArrowDown";
}

export function isArrowUp(event: KeyboardEvent): boolean {
    return event.key === "ArrowUp";
}

export function isTabKey(event: KeyboardEvent): boolean {
    return event.key === "Tab";
}

export function isBackspaceKey(event: KeyboardEvent): boolean {
    return event.key === "Backspace";
}
