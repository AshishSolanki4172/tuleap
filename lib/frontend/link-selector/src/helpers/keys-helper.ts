/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

export function isEscapeKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Escape" || keyboard_event.key === "Esc";
}

export function isEnterKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Enter";
}

export function isArrowDown(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "ArrowDown";
}

export function isArrowUp(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "ArrowUp";
}

export function isTabKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Tab";
}

export function isShiftKey(event: Event): event is KeyboardEvent {
    const keyboard_event = getKeyboardEvent(event);

    return keyboard_event.key === "Shift";
}

function getKeyboardEvent(event: Event): KeyboardEvent {
    if (!(event instanceof KeyboardEvent)) {
        throw new Error("Event is not a keyboard event");
    }

    return event;
}
