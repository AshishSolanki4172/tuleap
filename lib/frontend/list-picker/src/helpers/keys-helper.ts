/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

export function isEscapeKey(event: KeyboardEvent): boolean {
    return event.key === "Escape" || event.key === "Esc" || event.keyCode === 27;
}

export function isBackspaceKey(event: KeyboardEvent): boolean {
    return event.key === "Backspace" || event.keyCode === 8;
}

export function isEnterKey(event: KeyboardEvent): boolean {
    return event.key === "Enter" || event.keyCode === 13;
}

export function isArrowDown(event: KeyboardEvent): boolean {
    return event.key === "ArrowDown" || event.keyCode === 40;
}

export function isArrowUp(event: KeyboardEvent): boolean {
    return event.key === "ArrowUp" || event.keyCode === 38;
}

export function isTabKey(event: KeyboardEvent): boolean {
    return event.key === "Tab" || event.keyCode === 9;
}

export function isShiftKey(event: KeyboardEvent): boolean {
    return event.key === "Shift" || event.keyCode === 16;
}
