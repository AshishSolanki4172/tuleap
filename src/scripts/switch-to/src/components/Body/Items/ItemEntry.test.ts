/*
 * Copyright (c) Enalean, 2020 - present. All Rights Reserved.
 *
 * This file is a part of Tuleap.
 *
 * Tuleap is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * Tuleap is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Tuleap. If not, see <http://www.gnu.org/licenses/>.
 */

import { shallowMount } from "@vue/test-utils";
import type { QuickLink, ItemDefinition } from "../../../type";
import ItemEntry from "./ItemEntry.vue";
import { createTestingPinia } from "@pinia/testing";
import { defineStore } from "pinia";
import type { KeyboardNavigationState } from "../../../stores/type";
import { getGlobalTestOptions } from "../../../helpers/global-options-for-test";

describe("ItemEntry", () => {
    it("Displays a link with a cross ref", () => {
        const wrapper = shallowMount(ItemEntry, {
            props: {
                entry: {
                    icon_name: "fa-columns",
                    xref: "art #123",
                    title: "Lorem ipsum",
                    color_name: "lake-placid-blue",
                    quick_links: [] as QuickLink[],
                    project: {
                        label: "Guinea Pig",
                    },
                } as ItemDefinition,
                changeFocusCallback: jest.fn(),
            },
            global: getGlobalTestOptions(),
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("Displays a link with a quick links", () => {
        const wrapper = shallowMount(ItemEntry, {
            props: {
                entry: {
                    icon_name: "fa-columns",
                    xref: "art #123",
                    title: "Lorem ipsum",
                    color_name: "lake-placid-blue",
                    quick_links: [
                        { name: "TTM", icon_name: "fa-check", html_url: "/ttm" },
                        { name: "AD", icon_name: "fa-table", html_url: "/ad" },
                    ],
                    project: {
                        label: "Guinea Pig",
                    },
                } as ItemDefinition,
                changeFocusCallback: jest.fn(),
            },
            global: getGlobalTestOptions(),
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it("Displays a link with an icon", () => {
        const wrapper = shallowMount(ItemEntry, {
            props: {
                entry: {
                    icon_name: "fa-columns",
                    title: "Kanban",
                    color_name: "lake-placid-blue",
                    quick_links: [] as QuickLink[],
                    project: {
                        label: "Guinea Pig",
                    },
                } as ItemDefinition,
                changeFocusCallback: jest.fn(),
            },
            global: getGlobalTestOptions(),
        });

        expect(wrapper.element).toMatchSnapshot();
    });

    it.each(["ArrowUp", "ArrowDown", "ArrowLeft", "ArrowRight"])(
        "Changes the focus with arrow key %s",
        async (key) => {
            const entry = {
                icon_name: "fa-columns",
                title: "Kanban",
                color_name: "lake-placid-blue",
                quick_links: [] as QuickLink[],
                project: {
                    label: "Guinea Pig",
                },
            } as ItemDefinition;

            const changeFocusCallback = jest.fn();
            const wrapper = shallowMount(ItemEntry, {
                props: {
                    entry,
                    changeFocusCallback,
                },
                global: getGlobalTestOptions(),
            });

            await wrapper.find("[data-test=entry-link]").trigger("keydown", { key });

            expect(changeFocusCallback).toHaveBeenCalledWith({
                entry,
                key,
            });
        }
    );

    it("Forces the focus from the outside", async () => {
        const useKeyboardNavigationStore = defineStore("keyboard-navigation", {
            state: (): KeyboardNavigationState =>
                ({
                    programmatically_focused_element: null,
                } as KeyboardNavigationState),
        });

        const pinia = createTestingPinia();
        const store = useKeyboardNavigationStore(pinia);

        const entry = {
            icon_name: "fa-columns",
            title: "Kanban",
            color_name: "lake-placid-blue",
            quick_links: [] as QuickLink[],
            project: {
                label: "Guinea Pig",
            },
        } as ItemDefinition;

        const wrapper = shallowMount(ItemEntry, {
            props: {
                entry,
                changeFocusCallback: jest.fn(),
            },
            global: getGlobalTestOptions(pinia),
        });

        const link = wrapper.find("[data-test=entry-link]");
        if (!(link.element instanceof HTMLAnchorElement)) {
            throw Error("Unable to find the link");
        }

        const focus = jest.spyOn(link.element, "focus");

        await store.$patch({
            programmatically_focused_element: entry,
        });

        expect(focus).toHaveBeenCalled();
    });
});
