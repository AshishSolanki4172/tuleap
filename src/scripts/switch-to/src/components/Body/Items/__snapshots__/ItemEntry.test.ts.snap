// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemEntry Displays a link with a cross ref 1`] = `
<div
  class="switch-to-item-entry"
>
  <div
    class="switch-to-item-entry-with-links switch-to-item-entry-with-links-with-badge"
  >
    <a
      class="lake-placid-blue switch-to-item-entry-link"
      data-test="entry-link"
    >
      <i
        aria-hidden="true"
        class="fa fa-fw switch-to-item-entry-icon fa-columns"
      />
      <span
        class="switch-to-item-entry-badge cross-ref-badge cross-ref-badge-on-dark-background tlp-swatch-lake-placid-blue"
        data-test="item-xref"
      >
        <highlight-matching-text-stub
          text="art #123"
        />
      </span>
      <highlight-matching-text-stub
        class="switch-to-item-entry-label"
        data-test="item-title"
        text="Lorem ipsum"
      />
    </a>
    <!--v-if-->
  </div>
  <div
    class="switch-to-item-metadata"
  >
    <span
      class="switch-to-item-project"
    >
      Guinea Pig
    </span>
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;

exports[`ItemEntry Displays a link with a quick links 1`] = `
<div
  class="switch-to-item-entry"
>
  <div
    class="switch-to-item-entry-with-links switch-to-item-entry-with-links-with-badge"
  >
    <a
      class="lake-placid-blue switch-to-item-entry-link"
      data-test="entry-link"
    >
      <i
        aria-hidden="true"
        class="fa fa-fw switch-to-item-entry-icon fa-columns"
      />
      <span
        class="switch-to-item-entry-badge cross-ref-badge cross-ref-badge-on-dark-background tlp-swatch-lake-placid-blue"
        data-test="item-xref"
      >
        <highlight-matching-text-stub
          text="art #123"
        />
      </span>
      <highlight-matching-text-stub
        class="switch-to-item-entry-label"
        data-test="item-title"
        text="Lorem ipsum"
      />
    </a>
    <div
      class="switch-to-item-entry-quick-links"
    >
      
      <quick-link-stub
        class="switch-to-item-entry-quick-links-link"
        item="[object Object]"
        link="[object Object]"
      />
      <quick-link-stub
        class="switch-to-item-entry-quick-links-link"
        item="[object Object]"
        link="[object Object]"
      />
      
    </div>
  </div>
  <div
    class="switch-to-item-metadata"
  >
    <span
      class="switch-to-item-project"
    >
      Guinea Pig
    </span>
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;

exports[`ItemEntry Displays a link with an icon 1`] = `
<div
  class="switch-to-item-entry"
>
  <div
    class="switch-to-item-entry-with-links"
  >
    <a
      class="lake-placid-blue switch-to-item-entry-link"
      data-test="entry-link"
    >
      <i
        aria-hidden="true"
        class="fa fa-fw switch-to-item-entry-icon fa-columns"
      />
      <!--v-if-->
      <highlight-matching-text-stub
        class="switch-to-item-entry-label"
        data-test="item-title"
        text="Kanban"
      />
    </a>
    <!--v-if-->
  </div>
  <div
    class="switch-to-item-metadata"
  >
    <span
      class="switch-to-item-project"
    >
      Guinea Pig
    </span>
    <!--v-if-->
    <!--v-if-->
  </div>
</div>
`;
